All MediaRecorderPrivate should be running in the web content process.

jyavenard opened the pull request:
#### 72be95a9c52756b4412dacc18062120d5c784889
<pre>
All MediaRecorderPrivate should be running in the web content process.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281816">https://bugs.webkit.org/show_bug.cgi?id=281816</a>
<a href="https://rdar.apple.com/138239763">rdar://138239763</a>

Reviewed by Jer Noble.

We extend on the work started with the MediaRecorderPrivateWriterWebM introduced in 285061@main.
We add a new MediaRecorderPrivateEncoder that performs the encoding, encoded frames interleaving management
and writing through the specialised MediaRecorderPrivateWriter.

This new MediaRecorderPrivate can encode into any audio codec supported by AVFoundation (though only opus and AAC is enabled for now)
and all video codecs supported by VideoEncoder with webm (vp8, vp9) and mp4 (h264, h265, av1).

The code runs in its dedicated WorkQueue.

One of the key change which added a significant amount of technical complexity is that while we will generate 2s GOP
a call to MediaRecorder.requestData() will return fully playable media segment, formed such as each new segment starts
with a keyframe. The generated media is MSE compatible.
This way, the page fully controls the duration of the media segments generated by either setting
a time slice when creating a MediaRecorder or calling requestData() at regular interval.

When a new segment is required by the page, we flush all pending frames received to that point. As the audio encoder
has a certain latency, we may need to wait for a future audio frame to come which would allow the audio for the
current time to be available. When we are streaming, this is not a problem ; however if the recording has been paused
no audio will be coming. So instead will fully drain the encoder. As a side effect, resuming the recording
may cause an audible silence for the duration of the codec delay (55ms with Opus content).
In practice, this may not be noticeable as pausing and resuming a recording would produce noticeable
artifacts with audio.

Another technical change with the previous webm recorder is that the video time is now set to use
the start time of the last appended audio frame rather than the end. This too is also a consequence of the audio encoder
having a set latency and allowed for some simplifications in the code.

The new MediaRecorder add supports for the codec to be set and the mimetype for the content generated will
be mostly accurate (we use static values with common profiles rather than attempting to determing accurately
things like the codec level and profile.

The RemoteMediaRecorder being no longer used is fully removed along the related files.
The class MediaRecorderProvider and its overrides are removed.

Tests amended and corrected to reflect the more accurate behaviors.

* LayoutTests/http/wpt/mediarecorder/MediaRecorder-first-frame-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-first-frame.html:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-requestData.html:
* LayoutTests/http/wpt/mediarecorder/mimeType.html:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-mimetype-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-mimetype.html:
* LayoutTests/ipc/shared-video-frame-size-expected.txt: Removed.
* LayoutTests/ipc/shared-video-frame-size.html: Removed.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations: Amend test, was failing before, just failed a different way. There&apos;s no MediaStream support on WK1
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::isTypeSupported):
(WebCore::MediaRecorder::createMediaRecorderPrivate):
(WebCore::MediaRecorder::startRecording):
* Source/WebCore/Modules/mediarecorder/MediaRecorder.h:
* Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.cpp: Removed.
* Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.h: Removed.
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/Page.h:
(WebCore::Page::socketProvider):
(WebCore::Page::mediaRecorderProvider): Deleted.
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/platform/MediaSample.h:
(WebCore::MediaSample::presentationEndTime const):
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::toCMSampleBuffer):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
(WebCore::MediaRecorderPrivateAVFImpl::isTypeSupported):
(WebCore::MediaRecorderPrivateAVFImpl::create):
(WebCore::MediaRecorderPrivateAVFImpl::MediaRecorderPrivateAVFImpl):
(WebCore::MediaRecorderPrivateAVFImpl::~MediaRecorderPrivateAVFImpl):
(WebCore::MediaRecorderPrivateAVFImpl::startRecording):
(WebCore::MediaRecorderPrivateAVFImpl::videoFrameAvailable):
(WebCore::MediaRecorderPrivateAVFImpl::audioSamplesAvailable):
(WebCore::MediaRecorderPrivateAVFImpl::stopRecording):
(WebCore::MediaRecorderPrivateAVFImpl::pauseRecording):
(WebCore::MediaRecorderPrivateAVFImpl::resumeRecording):
(WebCore::MediaRecorderPrivateAVFImpl::fetchData):
(WebCore::MediaRecorderPrivateAVFImpl::mimeType const):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp: Added.
(WebCore::MediaRecorderPrivateEncoder::Listener::create):
(WebCore::MediaRecorderPrivateEncoder::Listener::Listener):
(WebCore::MediaRecorderPrivateEncoder::create):
(WebCore::MediaRecorderPrivateEncoder::queueSingleton):
(WebCore::MediaRecorderPrivateEncoder::MediaRecorderPrivateEncoder):
(WebCore::codecStringForMediaVideoCodecId):
(WebCore::MediaRecorderPrivateEncoder::compressedAudioOutputBufferCallback):
(WebCore::MediaRecorderPrivateEncoder::initialize):
(WebCore::MediaRecorderPrivateEncoder::generateMIMEType):
(WebCore::MediaRecorderPrivateEncoder::pause):
(WebCore::MediaRecorderPrivateEncoder::resume):
(WebCore::MediaRecorderPrivateEncoder::close):
(WebCore::MediaRecorderPrivateEncoder::mimeType const):
(WebCore::MediaRecorderPrivateEncoder::audioBitRate const):
(WebCore::MediaRecorderPrivateEncoder::videoBitRate const):
(WebCore::MediaRecorderPrivateEncoder::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::nextVideoFrameTime):
(WebCore::MediaRecorderPrivateEncoder::appendData):
(WebCore::MediaRecorderPrivateEncoder::flushDataBuffer):
(WebCore::MediaRecorderPrivateEncoder::hasMuxedDataSinceEndSegment const):
(WebCore::MediaRecorderPrivateEncoder::takeData):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::maybeStartWriter):
(WebCore::MediaRecorderPrivateEncoder::encodePendingVideoFrames):
(WebCore::MediaRecorderPrivateEncoder::processVideoEncoderActiveConfiguration):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::partiallyFlushEncodedQueues):
(WebCore::MediaRecorderPrivateEncoder::waitForMatchingAudio):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment):
(WebCore::MediaRecorderPrivateEncoder::flushAllEncodedQueues):
(WebCore::MediaRecorderPrivateEncoder::muxNextFrame):
(WebCore::MediaRecorderPrivateEncoder::stopRecording):
(WebCore::MediaRecorderPrivateEncoder::fetchData):
(WebCore::MediaRecorderPrivateEncoder::flushPendingData):
(WebCore::MediaRecorderPrivateEncoder::currentTime const):
(WebCore::MediaRecorderPrivateEncoder::currentEndTime const):
(WebCore::MediaRecorderPrivateEncoder::audioCompressor const):
(WebCore::MediaRecorderPrivateEncoder::listener):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h: Added.
(WebCore::MediaRecorderPrivateEncoder::hasAudio const):
(WebCore::MediaRecorderPrivateEncoder::hasVideo const):
(WebCore::MediaRecorderPrivateEncoder::WTF_GUARDED_BY_CAPABILITY):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
(WebCore::MediaRecorderPrivateWriter::create):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
(): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.h: Copied from Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm: Added.
(-[WebAVAssetWriterDelegate initWithWriter:]):
(-[WebAVAssetWriterDelegate assetWriter:didOutputSegmentData:segmentType:]):
(WebCore::MediaRecorderPrivateWriterAVFObjC::create):
(WebCore::MediaRecorderPrivateWriterAVFObjC::MediaRecorderPrivateWriterAVFObjC):
(WebCore::MediaRecorderPrivateWriterAVFObjC::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterAVFObjC::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterAVFObjC::allTracksAdded):
(WebCore::MediaRecorderPrivateWriterAVFObjC::muxFrame):
(WebCore::appendEndsPreviousSampleDurationMarker):
(WebCore::MediaRecorderPrivateWriterAVFObjC::forceNewSegment):
(WebCore::MediaRecorderPrivateWriterAVFObjC::close):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h: Removed.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm: Removed.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp: Added.
(WebCore::MediaRecorderPrivateWriterWebMDelegate::MediaRecorderPrivateWriterWebMDelegate):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addFrame):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::forceNewClusterOnNextFrame):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::finalize):
(WebCore::MediaRecorderPrivateWriterWebM::create):
(WebCore::mkvCodeIcForMediaVideoCodecId):
(WebCore::MediaRecorderPrivateWriterWebM::MediaRecorderPrivateWriterWebM):
(WebCore::MediaRecorderPrivateWriterWebM::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebM::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebM::muxFrame):
(WebCore::MediaRecorderPrivateWriterWebM::forceNewSegment):
(WebCore::MediaRecorderPrivateWriterWebM::close):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm: Removed.
* Source/WebCore/testing/Internals.cpp:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::allowsExitUnderMemoryPressure const):
(WebKit::GPUConnectionToWebProcess::dispatchMessage):
(WebKit::GPUConnectionToWebProcess::mediaRecorderManager): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.h: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.messages.in: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorderManager.cpp: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorderManager.h: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorderManager.messages.in: Removed.
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderIdentifier.h: Removed.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp: Removed.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.h: Removed.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderProvider.cpp: Removed.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderProvider.h: Removed.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: <a href="https://commits.webkit.org/285726@main">https://commits.webkit.org/285726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ca8a27f356fde6f4cadf2350480498b7f73402a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73640 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53069 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26451 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77945 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24878 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75755 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62202 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/854 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57907 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/16300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76707 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/48057 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63363 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38312 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/73148 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44676 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/20850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23211 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66777 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66389 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21198 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79528 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72897 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/957 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/428 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1099 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63373 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16196 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9404 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7585 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94679 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/921 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3671 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20819 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/950 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/937 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/956 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -0,0 +1,852 @@
+            m_videoCodec = 'vp08';
+        else if (codec == "opus"_s)
+            m_audioCodec = kAudioFormatOpus;
+        else if (!isWebM && (startsWithLettersIgnoringASCIICase(codec, "avc1"_s)))
+            m_videoCodec = kCMVideoCodecType_H264;
+        else if (!isWebM && (codec.startsWith("hev1."_s) || codec.startsWith("hvc1."_s)))
+            m_videoCodec = kCMVideoCodecType_HEVC;
```
This is fine for now, but I think we should parse these codec strings more fully so that we know, e.g., whether to use video- or full-range color.

jyavenard replied:
both the HEVC and H264 remote decoder default to use full-range.

jernoble commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -0,0 +1,852 @@
+            audioCompressor()->setBitsPerSecond(*options.audioBitsPerSecond);
+    }
+
+    if (hasAudio())
+        m_audioCodecMimeType = codecStringForMediaVideoCodecId(m_audioCodec);
+    if (hasVideo())
+        m_videoCodecMimeType = codecStringForMediaVideoCodecId(m_videoCodec);
```
...Because this strips out all the codec configuration information from the mime type string and replaces it with possibly conflicting encoding options.

jyavenard replied:
yes, but there's currently not way of knowing what came out of the encoder, for now there's very little thing we can request: bitrate, and even the colourspace is left blank on return..

We would need to scan the bitstream returned to accurately determine what was spit out. it's quite a process unfortunately.

We don't have code in place currently to do that for any of the codecs.

jernoble commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -0,0 +1,852 @@
+    // explicitly allow the following allocation(s).
+    DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
+
+    // FIXME: is the below insertion correct?
```
Do you mean "assertion"?

jernoble commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -0,0 +1,852 @@
+    if (m_isStopped)
+        return;
+
+    // Heap allocations are forbidden on the audio thread for performance reasons so we need to
+    // explicitly allow the following allocation(s).
+    DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
```
If heap allocations are forbidden, shouldn't we `dispatch_async` to a non-realtime thread to do this work?

jyavenard replied:
This is what we are doing here.

The audio `PlatformAudioData` buffer we are receiving isn't a ref-counted object, it needs to be copied before hand into an object that is . That's the job of `createAudioSampleBuffer` 
the dispatch async itself is is also causing the issue, there's an allocation required to build the function given to the dispatch. And the AudioSampleBufferConverter itself dispatch to a non-realtime thread already.

this code pattern is present in almost all audio listener, so I don't think one more will make a difference.

jernoble commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -0,0 +1,852 @@
+{
+    m_inflightVideoFrames++;
+
+    queue().dispatch([weakThis = ThreadSafeWeakPtr { *this }, this, frame = Ref { frame }, sampleCount = static_cast<size_t>(m_lastEnqueuedAudioSampleCount), samplingRate = static_cast<uint32_t>(m_currentAudioSamplingRate)]() mutable {
```
It seems like we have a queue handy....

jernoble commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -0,0 +1,852 @@
+    }
+    m_encodedVideoFrames.append(MediaSampleAVFObjC::create(sample->get(), m_videoTrackIndex));
+    Ref sampleObjC = m_encodedVideoFrames.last();
+    ASSERT(frame.isKeyFrame == bool(sampleObjC->flags() & MediaSample::SampleFlags::IsSync));
```
This can just be: 

```c++
    ASSERT(frame.isKeyFrame == sampleObjC->isSync());
```

jernoble commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2024 Apple Inc. All rights reserved.
+ * Copyright (C) 2018 Apple Inc. All rights reserved.
```
This seems wrong.

jernoble commented on Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
```diff
@@ -32,7 +32,8 @@
 
 typedef struct opaqueCMSampleBuffer *CMSampleBufferRef;
-typedef struct OpaqueAudioConverter* AudioConverterRef;
+typedef struct OpaqueAudioConverter *AudioConverterRef;
```
AudioConverter is not an Objective-C type, so the indirection symbol goes next to the type.

jernoble commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
```diff
@@ -0,0 +1,217 @@
+    m_audioTrackIndex = ++m_currentTrackIndex;
+
+    return m_audioTrackIndex;
+}
```
It seems like this function should return std::optional<uint8_t>, since "0" is a perfectly valid track ID otherwise.

jyavenard replied:
not for the muxer, a track of 0 is invalid and 0 is the code returned by the mkvmuxer if it failed to add a track.
So we always use track ID 1 or 2

jyavenard replied:
I've changed it to std::optional<uint8_t>

jernoble commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
```diff
@@ -47,15 +49,61 @@ namespace WebCore {
+
+    if (equalLettersIgnoringASCIICase(containerType, "audio/mp4"_s) || equalLettersIgnoringASCIICase(containerType, "video/mp4"_s)) {
+        for (auto& codec : mimeType.codecs()) {
+            // FIXME: We should further validate parameters.
+            if (!startsWithLettersIgnoringASCIICase(codec, "avc1"_s)
```
We should probably re-use the same path for MediaRecorder as we do for MediaCapabilities; I don't believe we have a good implementation for the encoding side of that API, so that would remain a "FIXME".

jyavenard committed:
All MediaRecorderPrivate should be running in the web content process.
https://bugs.webkit.org/show_bug.cgi?id=281816
rdar://138239763

Reviewed by Jer Noble.

We extend on the work started with the MediaRecorderPrivateWriterWebM introduced in 285061@main.
We add a new MediaRecorderPrivateEncoder that performs the encoding, encoded frames interleaving management
and writing through the specialised MediaRecorderPrivateWriter.

This new MediaRecorderPrivate can encode into any audio codec supported by AVFoundation (though only opus and AAC is enabled for now)
and all video codecs supported by VideoEncoder with webm (vp8, vp9) and mp4 (h264, h265, av1).

The code runs in its dedicated WorkQueue.

One of the key change which added a significant amount of technical complexity is that while we will generate 2s GOP
a call to MediaRecorder.requestData() will return fully playable media segment, formed such as each new segment starts
with a keyframe. The generated media is MSE compatible.
This way, the page fully controls the duration of the media segments generated by either setting
a time slice when creating a MediaRecorder or calling requestData() at regular interval.

When a new segment is required by the page, we flush all pending frames received to that point. As the audio encoder
has a certain latency, we may need to wait for a future audio frame to come which would allow the audio for the
current time to be available. When we are streaming, this is not a problem ; however if the recording has been paused
no audio will be coming. So instead will fully drain the encoder. As a side effect, resuming the recording
may cause an audible silence for the duration of the codec delay (55ms with Opus content).
In practice, this may not be noticeable as pausing and resuming a recording would produce noticeable
artifacts with audio.

Another technical change with the previous webm recorder is that the video time is now set to use
the start time of the last appended audio frame rather than the end. This too is also a consequence of the audio encoder
having a set latency and allowed for some simplifications in the code.

The new MediaRecorder add supports for the codec to be set and the mimetype for the content generated will
be mostly accurate (we use static values with common profiles rather than attempting to determing accurately
things like the codec level and profile.

The RemoteMediaRecorder being no longer used is fully removed along the related files.
The class MediaRecorderProvider and its overrides are removed.

Tests amended and corrected to reflect the more accurate behaviors.

* LayoutTests/http/wpt/mediarecorder/MediaRecorder-first-frame-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-first-frame.html:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-requestData.html:
* LayoutTests/http/wpt/mediarecorder/mimeType.html:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-mimetype-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-mimetype.html:
* LayoutTests/ipc/shared-video-frame-size-expected.txt: Removed.
* LayoutTests/ipc/shared-video-frame-size.html: Removed.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations: Amend test, was failing before, just failed a different way. There's no MediaStream support on WK1
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::isTypeSupported):
(WebCore::MediaRecorder::createMediaRecorderPrivate):
(WebCore::MediaRecorder::startRecording):
* Source/WebCore/Modules/mediarecorder/MediaRecorder.h:
* Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.cpp: Removed.
* Source/WebCore/Modules/mediarecorder/MediaRecorderProvider.h: Removed.
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/Page.h:
(WebCore::Page::socketProvider):
(WebCore::Page::mediaRecorderProvider): Deleted.
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/platform/MediaSample.h:
(WebCore::MediaSample::presentationEndTime const):
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::toCMSampleBuffer):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
(WebCore::MediaRecorderPrivateAVFImpl::isTypeSupported):
(WebCore::MediaRecorderPrivateAVFImpl::create):
(WebCore::MediaRecorderPrivateAVFImpl::MediaRecorderPrivateAVFImpl):
(WebCore::MediaRecorderPrivateAVFImpl::~MediaRecorderPrivateAVFImpl):
(WebCore::MediaRecorderPrivateAVFImpl::startRecording):
(WebCore::MediaRecorderPrivateAVFImpl::videoFrameAvailable):
(WebCore::MediaRecorderPrivateAVFImpl::audioSamplesAvailable):
(WebCore::MediaRecorderPrivateAVFImpl::stopRecording):
(WebCore::MediaRecorderPrivateAVFImpl::pauseRecording):
(WebCore::MediaRecorderPrivateAVFImpl::resumeRecording):
(WebCore::MediaRecorderPrivateAVFImpl::fetchData):
(WebCore::MediaRecorderPrivateAVFImpl::mimeType const):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp: Added.
(WebCore::MediaRecorderPrivateEncoder::Listener::create):
(WebCore::MediaRecorderPrivateEncoder::Listener::Listener):
(WebCore::MediaRecorderPrivateEncoder::create):
(WebCore::MediaRecorderPrivateEncoder::queueSingleton):
(WebCore::MediaRecorderPrivateEncoder::MediaRecorderPrivateEncoder):
(WebCore::codecStringForMediaVideoCodecId):
(WebCore::MediaRecorderPrivateEncoder::compressedAudioOutputBufferCallback):
(WebCore::MediaRecorderPrivateEncoder::initialize):
(WebCore::MediaRecorderPrivateEncoder::generateMIMEType):
(WebCore::MediaRecorderPrivateEncoder::pause):
(WebCore::MediaRecorderPrivateEncoder::resume):
(WebCore::MediaRecorderPrivateEncoder::close):
(WebCore::MediaRecorderPrivateEncoder::mimeType const):
(WebCore::MediaRecorderPrivateEncoder::audioBitRate const):
(WebCore::MediaRecorderPrivateEncoder::videoBitRate const):
(WebCore::MediaRecorderPrivateEncoder::appendAudioSampleBuffer):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::nextVideoFrameTime):
(WebCore::MediaRecorderPrivateEncoder::appendData):
(WebCore::MediaRecorderPrivateEncoder::flushDataBuffer):
(WebCore::MediaRecorderPrivateEncoder::hasMuxedDataSinceEndSegment const):
(WebCore::MediaRecorderPrivateEncoder::takeData):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::maybeStartWriter):
(WebCore::MediaRecorderPrivateEncoder::encodePendingVideoFrames):
(WebCore::MediaRecorderPrivateEncoder::processVideoEncoderActiveConfiguration):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::partiallyFlushEncodedQueues):
(WebCore::MediaRecorderPrivateEncoder::waitForMatchingAudio):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment):
(WebCore::MediaRecorderPrivateEncoder::flushAllEncodedQueues):
(WebCore::MediaRecorderPrivateEncoder::muxNextFrame):
(WebCore::MediaRecorderPrivateEncoder::stopRecording):
(WebCore::MediaRecorderPrivateEncoder::fetchData):
(WebCore::MediaRecorderPrivateEncoder::flushPendingData):
(WebCore::MediaRecorderPrivateEncoder::currentTime const):
(WebCore::MediaRecorderPrivateEncoder::currentEndTime const):
(WebCore::MediaRecorderPrivateEncoder::audioCompressor const):
(WebCore::MediaRecorderPrivateEncoder::listener):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h: Added.
(WebCore::MediaRecorderPrivateEncoder::hasAudio const):
(WebCore::MediaRecorderPrivateEncoder::hasVideo const):
(WebCore::MediaRecorderPrivateEncoder::WTF_GUARDED_BY_CAPABILITY):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
(WebCore::MediaRecorderPrivateWriter::create):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
(): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.h: Copied from Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm: Added.
(-[WebAVAssetWriterDelegate initWithWriter:]):
(-[WebAVAssetWriterDelegate assetWriter:didOutputSegmentData:segmentType:]):
(WebCore::MediaRecorderPrivateWriterAVFObjC::create):
(WebCore::MediaRecorderPrivateWriterAVFObjC::MediaRecorderPrivateWriterAVFObjC):
(WebCore::MediaRecorderPrivateWriterAVFObjC::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterAVFObjC::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterAVFObjC::allTracksAdded):
(WebCore::MediaRecorderPrivateWriterAVFObjC::muxFrame):
(WebCore::appendEndsPreviousSampleDurationMarker):
(WebCore::MediaRecorderPrivateWriterAVFObjC::forceNewSegment):
(WebCore::MediaRecorderPrivateWriterAVFObjC::close):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.h: Removed.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm: Removed.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp: Added.
(WebCore::MediaRecorderPrivateWriterWebMDelegate::MediaRecorderPrivateWriterWebMDelegate):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addFrame):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::forceNewClusterOnNextFrame):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::finalize):
(WebCore::MediaRecorderPrivateWriterWebM::create):
(WebCore::mkvCodeIcForMediaVideoCodecId):
(WebCore::MediaRecorderPrivateWriterWebM::MediaRecorderPrivateWriterWebM):
(WebCore::MediaRecorderPrivateWriterWebM::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebM::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebM::muxFrame):
(WebCore::MediaRecorderPrivateWriterWebM::forceNewSegment):
(WebCore::MediaRecorderPrivateWriterWebM::close):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm: Removed.
* Source/WebCore/testing/Internals.cpp:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::allowsExitUnderMemoryPressure const):
(WebKit::GPUConnectionToWebProcess::dispatchMessage):
(WebKit::GPUConnectionToWebProcess::mediaRecorderManager): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.h: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.messages.in: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorderManager.cpp: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorderManager.h: Removed.
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorderManager.messages.in: Removed.
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderIdentifier.h: Removed.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp: Removed.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.h: Removed.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderProvider.cpp: Removed.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderProvider.h: Removed.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

Canonical link: https://commits.webkit.org/285726@main
